transmogrifier/exceptions.py

```diff
@@ -0,0 +1,21 @@
+class DeletedRecordEvent(Exception):  # noqa: N818
+    """Exception raised for records with a deleted status.
+
+    Attributes:
+        timdex_record_id: The TIMDEX record ID (not the source record ID) for the record.
+    """
+
+    def __init__(self, timdex_record_id: str) -> None:
+        self.timdex_record_id = timdex_record_id
+
+
+class SkippedRecordEvent(Exception):  # noqa: N818
+    """Exception raised for records that should be skipped.
+
+    Attributes:
+        source_record_id: The ID for the source record.
+    """
+
+    def __init__(self, source_record_id: str | None, message: str) -> None:
+        self.source_record_id = source_record_id
+        self.message = message
```

Comment:
I think there is some nuance to consider here, but feels like the time to think about these exceptions as a) they are now going into `main`, and b) might be fairly central to new orchestration patterns when implemented.

I would propose two things:
1. We can instantiate these exceptions just with a string message if we like, just like normal exceptions, and have this represented when you stringify the exception object
2. All _extra_ properties are optional, unless we decide they should be required

Example of how this could be done:
```python
class SkippedRecordEvent(Exception):
    """"""
    def __init__(self, message:str|None=None, source_record_id:str|None=None):
        super().__init__(message) # calls base Exception constructor
        self.source_record_id = source_record_id
```

By default, the base `Exception` class is expecting a positional argument `message` that is then used in the `__str__` method to stringify it.

This allows some behavior like this:
```
# acts like other exceptions, allows string message as first and only argument
In [19]: e = SkippedRecordEvent("Hello world!")

In [20]: str(e)
Out[20]: 'Hello world!'

In [21]: e
Out[21]: __main__.SkippedRecordEvent('Hello world!')

In [22]: e.source_record_id
# None passed

# also supports extra properties if defined
In [23]: e2 = SkippedRecordEvent("Hello world!", source_record_id="acb123")

In [24]: str(e2)
Out[24]: 'Hello world!'

# source_record_id also present if we want to use it
In [25]: e2.source_record_id
Out[25]: 'acb123'

# not ideal, but like default exceptions, can even raise it without messages
In [26]: e3 = SkippedRecordEvent()

In [27]: str(e3)
Out[27]: 'None' # string of None

In [28]: e3.source_record_id
# also None
```

Two additional considerations, maybe now, maybe later:
1. Does this have implications for `DeletedRecordEvent`?  Do we want to update patterns so it has similar default, predictable behavior?
2. Do we want more granular exception classes like `InvalidRecordError` that extend `SkippedRecordEvent`, such that we could raise those more meaningful, granular exceptions from code, but still have a single `except SkippedRecordEvent: ...` logic at the higher level to group skipping behavior for all those child Exception types?

Both seem like they could wait.  And for that matter, the suggestions above could too.  But felt like a good time to raise them.

Reply:
Agree on the proposed exception formatting!

Re: additional considerations, those feel better to address during the full orchestration discussion. It doesn't feel like we have full use cases for either yet (correct me if I'm wrong)but I would like to discuss both in the more comprehensive context of the orchestration refactor.

If that sounds OK, we can create an issue with those 2 points under something like `Exception refactoring` and as related items come up they could be added as comments. Open to other ways of doing it though!

Reply:
Sounds good to me.  If `SkippedRecordEvent` is updated now to allow for default-y Exception behavior discussed above, then the rest feels like something we could suss out in future work.  There aren't that many instances of raising these from code, so would be easy to `ctrl + f` them and then think about what new exceptions might make sense to raise there.

Reply:
New commit pushed!